src/callbacks_step/ape_euler_coupling.jl

```diff
@@ -0,0 +1,111 @@
+# TODO: Allow kwargs that are later used when constructing `integrator_euler`?
+mutable struct ApeEulerCouplingCallback{RealT<:Real}
+  stepsize_callback_ape::StepsizeCallback{RealT}
+  stepsize_callback_euler::StepsizeCallback{RealT}
+  averaging_callback::DiscreteCallback{<:Any, <:AveragingCallback}
+  integrator_euler::Union{Nothing, AbstractODEIntegrator}
+end
+
+
+function Base.show(io::IO, cb::DiscreteCallback{<:Any, <:ApeEulerCouplingCallback})
+
+end
+
+function Base.show(io::IO, ::MIME"text/plain", cb::DiscreteCallback{<:Any, <:ApeEulerCouplingCallback})
+
+end
+
+
+function ApeEulerCouplingCallback(cfl_ape::Real, cfl_euler::Real,
+                                  averaging_callback::DiscreteCallback{<:Any, <:AveragingCallback})
+
+  ape_euler_coupling = ApeEulerCouplingCallback{typeof(cfl_ape)}(StepsizeCallback(cfl_ape),
+                                                                 StepsizeCallback(cfl_euler),
+                                                                 averaging_callback, nothing)
+  condition = (u, t, integrator) -> true
+
+  return DiscreteCallback(condition, ape_euler_coupling, save_positions=(false, false),
+                          initialize=initialize!)
+end
+
+function initialize!(cb::DiscreteCallback{Condition,Affect!}, u_ode, t, integrator) where {Condition, Affect!<:ApeEulerCouplingCallback}
+  ape_euler_coupling = cb.affect!
+  semi = integrator.p
+  @unpack semi_ape, semi_euler = semi
+
+  # Set up ODE Integrator for Euler equations
+  tspan = integrator.sol.prob.tspan
+  alg = integrator.alg
+
+  ode_euler = semidiscretize(semi_euler, tspan)
+  ape_euler_coupling.integrator_euler = init(ode_euler, alg, save_everystep=false, dt=1.0) # dt will be overwritten
+
+  # Set mean values in u_ode according to `AveragingCallback`
+  u_ape = wrap_array(u_ode, semi_ape)
+  mesh, equations, solver, cache = mesh_equations_solver_cache(semi)
+  @unpack mean_values = ape_euler_coupling.averaging_callback.affect!
+  @views @. u_ape[4:5, .., :] = mean_values.v_mean
+  @views @. u_ape[6, .., :] = mean_values.c_mean
+  @views @. u_ape[7, .., :] = mean_values.rho_mean
+
+  # Calculate gradient of squared mean speed of sound for the q_cons source term
+  calc_gradient_c_mean_square!(semi.cache.grad_c_mean_sq, u_ape, mesh, equations, solver, cache)
```

Comment:
Is there a way to separately time the conservation source term? I have the feeling that we have to start rewriting the APE system such that we can get rid of it for good, essentially using the current variables as primitive variables and `p/c^2`, `u`, `v`, as conservative.

Reply:
Thanks!